pkg/provider/oidc.go

```diff
@@ -0,0 +1,195 @@
+				return
+			}
+
+			pi := info.ProviderInfoFromContext(ctx)
+			resp, err := checkIntrospectWithCached(ctx, resourceServer, token, pi.IntrospectionResponse)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusForbidden)
+				return
+			}
+			ctx = pi.SetIntrospectionResponse(resp).IntoContext(ctx)
+
+			if err := checkAuthorization(pi.IntrospectionResponse, requestedClaim, requestedValue); err != nil {
+				http.Error(w, err.Error(), http.StatusForbidden)
+				return
+			}
+			next.ServeHTTP(w, req.WithContext(ctx))
+		})
+	}
+}
+
+func oidcAuthenticatedUnaryInterceptor(resourceServer rs.ResourceServer) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, sInfo *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
+		token, err := checkOIDCToken(metautils.ExtractIncoming(ctx).Get("authorization"))
+		if err != nil {
+			return nil, status.Error(codes.Unauthenticated, err.Error())
+		}
+
+		pi := info.ProviderInfoFromContext(ctx)
+		iResp, err := checkIntrospectWithCached(ctx, resourceServer, token, pi.IntrospectionResponse)
+		if err != nil {
+			return nil, status.Error(codes.PermissionDenied, err.Error())
+		}
+		ctx = pi.SetIntrospectionResponse(iResp).IntoContext(ctx)
+
+		return handler(ctx, req)
+	}
+}
+
+type serverStream struct {
+	grpc.ServerStream
+	ctx context.Context
+}
+
+func (s *serverStream) Context() context.Context {
+	return s.ctx
+}
+
+func oidcAuthenticatedStreamInterceptor(resourceServer rs.ResourceServer) grpc.StreamServerInterceptor {
+	return func(srv interface{}, stream grpc.ServerStream, sInfo *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		ctx := stream.Context()
+
+		token, err := checkOIDCToken(metautils.ExtractIncoming(ctx).Get("authorization"))
+		if err != nil {
+			return status.Error(codes.Unauthenticated, err.Error())
+		}
+
+		pi := info.ProviderInfoFromContext(ctx)
+		resp, err := checkIntrospectWithCached(ctx, resourceServer, token, pi.IntrospectionResponse)
+		if err != nil {
+			return status.Error(codes.PermissionDenied, err.Error())
+		}
+		ctx = pi.SetIntrospectionResponse(resp).IntoContext(ctx)
+
+		return handler(srv, &serverStream{stream, ctx})
+	}
+}
+
+func checkOIDCTokenFromRequest(r *http.Request) (string, error) {
+	auth := r.Header.Get("authorization")
+	if auth == "" {
+		return "", ErrMissingHeader
+	}
+	return checkOIDCToken(auth)
+}
+
+func checkOIDCToken(auth string) (string, error) {
+	if !strings.HasPrefix(auth, oidc.PrefixBearer) {
+		return "", ErrInvalidHeader
+	}
+	return strings.TrimPrefix(auth, oidc.PrefixBearer), nil
+}
+
+func checkIntrospect(ctx context.Context, resourceServer rs.ResourceServer, token string) (*oidc.IntrospectionResponse, error) {
+	resp, err := rs.Introspect[*oidc.IntrospectionResponse](ctx, resourceServer, token)
+	if err != nil {
+		return nil, ErrInvalidToken
+	}
+	if !resp.Active {
+		return nil, ErrInvalidToken
+	}
+	return resp, nil
+}
```

Comment:
as mentioned before, i'd make that generic so that developers can handle their custom claims (if needed)